pinot-core/src/main/java/org/apache/pinot/core/transport/ServerChannels.java

```diff
@@ -139,22 +143,32 @@ private void attachSSLHandler(SocketChannel ch) {
       }
     }
 
-    synchronized void sendRequest(String rawTableName, AsyncQueryResponse asyncQueryResponse,
-        ServerRoutingInstance serverRoutingInstance, InstanceRequest instanceRequest)
+    private void sendRequestOrTimeOut(String rawTableName, AsyncQueryResponse asyncQueryResponse,
+        ServerRoutingInstance serverRoutingInstance, byte[] requestBytes, long timeoutMs)
+        throws Exception {
+      if (_channelLock.tryLock(timeoutMs, TimeUnit.MILLISECONDS)) {
+        try {
+          sendRequest(rawTableName, asyncQueryResponse, serverRoutingInstance, requestBytes);
+        } finally {
+          _channelLock.unlock();
+        }
+      }
```

Comment:
(major) Throw timeout exception if it fails to acquire the lock

Reply:
The `lock.tryLock()` itself will throw an `InterruptedException` if timeout.

Reply:
Jackie is right, sorry I missed this.

```
     * @return {@code true} if the lock was free and was acquired by the
     *         current thread, or the lock was already held by the current
     *         thread; and {@code false} if the waiting time elapsed before
     *         the lock could be acquired
     * @throws InterruptedException if the current thread is interrupted
     * @throws NullPointerException if the time unit is null
     */
```

Reply:
I missed that too 😂 . Done